contracts/governance/BaseWeightedMultisig.sol

```diff
@@ -0,0 +1,206 @@
+        return _baseWeightedStorage().signersEpochForHash[signerHash];
+    }
+
+    /*
+     * @notice This function takes messageHash and proof data and reverts if proof is invalid
+     * @param messageHash The hash of the message that was signed
+     * @param weightedSigners The weighted signers data
+     * @param signatures The signatures data
+     * @return isLatestSigners True if provided signers are the current ones
+     */
+    function validateProof(bytes32 messageHash, bytes calldata proof) public view returns (bool isLatestSigners) {
+        WeightedMultisigStorage storage slot = _baseWeightedStorage();
+
+        if (proof.length < 32) revert InvalidSigners();
+
+        WeightedSigners memory signers;
+        bytes[] memory signatures;
+
+        (signers.accounts, signers.weights, signers.threshold, signatures) = abi.decode(
+            proof,
+            (address[], uint256[], uint256, bytes[])
+        );
+
+        bytes32 signersHash = keccak256(abi.encode(signers.accounts, signers.weights, signers.threshold));
+        uint256 epoch = slot.signersEpochForHash[signersHash];
+        uint256 currentEpoch = slot.currentSignersEpoch;
+
+        isLatestSigners = epoch == currentEpoch;
+
+        if (signatures.length == 0) revert MalformedSignatures();
+        if (epoch == 0 || currentEpoch - epoch > OLD_SIGNERS_RETENTION) revert InvalidSigners();
+
+        _validateSignatures(messageHash, signers, signatures);
+    }
+
+    /*************************\
+    |* Integration Functions *|
+    \*************************/
+
+    /*
+     * @notice This function rotates the current signers with a new set of signers
+     * @param newWeightedSigners The new weighted signers data
+     */
+    function _rotateSigners(WeightedSigners memory newSigners) internal {
+        WeightedMultisigStorage storage slot = _baseWeightedStorage();
+
+        uint256 length = newSigners.accounts.length;
+        uint256 totalWeight;
+
+        // signers must be sorted binary or alphabetically in lower case
+        if (newSigners.accounts.length == 0 || !_isSortedAscAndContainsNoDuplicate(newSigners.accounts))
+            revert InvalidSigners();
+
+        if (newSigners.weights.length != length) revert InvalidWeights();
+
+        for (uint256 i; i < length; ++i) {
+            uint256 weight = newSigners.weights[i];
+
+            if (weight == 0) revert InvalidWeights();
+
+            totalWeight = totalWeight + weight;
+        }
+
+        if (newSigners.threshold == 0 || totalWeight < newSigners.threshold) revert InvalidThreshold();
+
+        bytes32 newSignersHash = keccak256(abi.encode(newSigners.accounts, newSigners.weights, newSigners.threshold));
+
+        uint256 epoch = slot.currentSignersEpoch + 1;
+        // slither-disable-next-line costly-loop
+        slot.currentSignersEpoch = epoch;
+        slot.hashForSignersEpoch[epoch] = newSignersHash;
+        slot.signersEpochForHash[newSignersHash] = epoch;
```

Comment:
Add a comment indicating that we overwrite the latest epoch for a signerHash if it's repeated